File: Greedy_algorithm/greedy.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define debug 0
class Route
{
public:
    ll Q;
    double time;
    ll id;
    Route(ll id, ll Q, double time)
    {
        this->id = id;
        this->Q = Q;
        this->time = time;
    }

    Route()
    {
        this->Q = 0;
        this->time = 0;
    }

    friend bool operator<(const Route &a, const Route &b)
    {
        // the route with the highest time is first
        return a.time > b.time;
    }
};

class Agent
{
public:
    ll id;
    double sum_time;
    Agent(ll id)
    {
        this->id = id;
        this->sum_time = 0;
    }

    Agent()
    {
        this->id = 0;
        this->sum_time = 0;
    }

    void add_time(double time)
    {
        if (debug)
            cout << "time adder: agent " << this->id << " of time " << this->sum_time << " increases by " << time << endl;
        this->sum_time += time;
        this->sum_time = round(this->sum_time * 1000) / 1000;
    }

    friend bool operator<(const Agent &a, const Agent &b)
    {
        // the agent with the highest sum of times is first
        return a.sum_time < b.sum_time;
    }
};

vector<Agent> agents;
vector<Route> routes;
vector<vector<ll>> history; // history of the agents' positions day by day
ll Q; // number of agents
ll K; // number of routes

void read_input()
{
    cin >> Q;
    cin >> K;
    agents.clear();
    agents.resize(Q);
    routes.clear();
    routes.resize(K);

    ll i = 0;
    for (int j = 0; j < K; j++)
    {
        ll q;
        double t;
        cin >> q >> t;
        routes[j] = Route(j,q, t);
    }
    for (int i = 0; i < Q; i++)
    {
        agents[i] = Agent(i);
    }

    history.clear();

    if (debug)
    {
        cout << "Agents: " << endl;
        for (int i = 0; i < agents.size(); i++)
        {
            cout << agents[i].id << " ";
        }
        cout << endl;

        cout << "Routes: " << endl;
        for (int i = 0; i < routes.size(); i++)
        {
            cout << routes[i].Q << " " << routes[i].time << endl;
        }
    }
}

void prepare_routes()
{
    double min_time = min_element(routes.begin(), routes.end(), [](Route a, Route b) { return a.time < b.time; })->time;
    for (auto &route : routes)
    {
        route.time -= min_time;
    }

    sort(routes.begin(), routes.end());

    if (debug)
    {
        cout << "Sorted routes: " << endl;
        for (int i = 0; i < routes.size(); i++)
        {
            cout << routes[i].Q << " " << routes[i].time << endl;
        }
    }
}

vector<ll> assign_routes(vector<Route> routes, vector<Agent> agents)
{
    // given sorted routes and agents, assign the routes to the agents in a greedy way
    // i.e assign the best route to q agents with the highest sum of times
    // returns the vector of the assigned routes for each agent
    
    vector<ll> assigned_routes(agents.size(), 0);
    ll idx = 0;
    for (int i=0;i < routes.size(); i++)
    {
        for (int j=0;j<routes[i].Q;j++)
        {
            assigned_routes[idx] = i;
            idx++;
        }
    }

    return assigned_routes;
}

void simulation_step()
{
    vector<ll> assigned_routes = assign_routes(routes, agents);


    for (int i = 0; i < agents.size(); i++)
    {
        agents[i].add_time(routes[assigned_routes[i]].time);
    }

    vector<ll> agent_history(agents.size(), 0);

    for (int i = 0; i < agents.size(); i++)
    {
        agent_history[agents[i].id] = assigned_routes[i];
    }

    history.push_back(agent_history);

    sort(agents.begin(), agents.end()); // sort the agents by the sum of times
   
}

void step_stats()
{
    cout << "Agent stats: " << endl;
    for (int i = 0; i < agents.size(); i++)
    {
        cout << "Agent " << agents[i].id << " has time " << agents[i].sum_time << endl;
    }
}

bool check_conv(double eps, ll cntr)
{
    double min_time = min_element(agents.begin(), agents.end(), [](Agent a, Agent b) { return a.sum_time < b.sum_time; })->sum_time;
    double max_time = max_element(agents.begin(), agents.end(), [](Agent a, Agent b) { return a.sum_time < b.sum_time; })->sum_time;
    double max_diff = max_time - min_time;
    if (max_diff > eps)
    {
        cout << "Not converged yet! Max diff: " << max_diff << endl;
        return false;
    }
    cout << "Converged in " << cntr << " steps!" << endl;
    cout << "Agents time: " << endl;
    for (int i = 0; i < agents.size(); i++)
    {
        cout << agents[i].sum_time << " ";
    }
    return true;
}

void simul_stats()
{
    cout << "Simulation stats: " << endl;
    
    vector<vector<ll>> agent_stats(agents.size(), vector<ll>(routes.size(), 0));

    for (int i = 0; i < history.size(); i++)
    {
        for (int j = 0; j < history[i].size(); j++)
        {
            agent_stats[j][history[i][j]]++;
        }
    }

    for (int i = 0; i < agent_stats.size(); i++)
    {
        cout << "Agent " << i << " stats: " << endl;
        for (int j = 0; j < agent_stats[i].size(); j++)
        {
            cout << agent_stats[i][j] << " ";
        }
        cout << endl;
    }
}

int main()
{
    read_input();

    prepare_routes();
    /*
    This greedy algorithm is a simple implementation of the Wardropian principle.
    The algorithm is not guaranteed to converge to the optimal solution, but it is a good starting point.
    What it does is to simulate road assignment to agents in a greedy way - at each step, the algorithm assigns the best route to the agent that has the highest sum of times.
    The algorithm stops when the convergence condition is met, i.e., when the difference between the sum of times of the best and the second best agents is less than eps.
    */

    bool converge = false;
    ll cntr = 0;
    while (!converge)
    {
        cntr++;
        simulation_step();

        if (cntr % 100 == 0)
        {
            step_stats();
        }

        if(check_conv(0, cntr))
        {
            break;
        }

        if (cntr > 10000)
        {
            cout << "Did not converge!" << endl;
            break;
        }
    }
    
    simul_stats();
}

File: Greedy_algorithm/greedy.py
import math
import csv
from functools import reduce
from utils import Route, Agent



def prep_data(route_data):
    '''
    Prepare the data for the simulation - round q to the nearest integer
    
    Parameters:
        route_data (list of tuple): Each tuple represents a route as (flow, time), where flow is the number of agents that can take this route, and time is the associated time.
        recommended to use the Route object from utils.py
        
    Returns:
        tuple: A tuple containing the total number of agents needed and the route data.
    '''
    route_data = [(round(r.flow), r.time) for r in route_data]
    # calculate the total number of agents needed
    total_agents = sum(q for q, _ in route_data)
    return total_agents, route_data

def run_simulation(num_agents, route_data, thresholds=[], max_steps=10000, eps=0, debug=False):
    """
    Runs the simulation for a given number of agents and route data.

    Parameters:
        num_agents (int): The number of agents.
        route_data (list of Route object): Each tuple represents a route as (flow, time), where flow is the number of agents that can take this route, and time is the associated time.
        thresholds (list of float): A list of convergence thresholds other than absolute convergence. Default is an empty list.
        max_steps (int): The maximum number of simulation steps. Default is 10,000.

    Returns:
        dict: A dictionary containing the history of agent positions, mean times, variances and fairness for each step.
    """
    agents = [Agent(i) for i in range(num_agents)]
    routes = [Route(j, q, t) for j, (q, t) in enumerate(route_data)]
    history = []
    mean = []
    variance = []
    fairness = []
    fairness_norm = []
    almost_convergence = [-1] * len(thresholds)

    def prepare_routes():
        # min_time = min(routes, key=lambda r: r.time).time
        # for route in routes:
        #     route.time -= min_time
        routes.sort()

    def assign_routes(routes, agents):
        assigned_routes = [0] * len(agents)
        idx = 0
        for i in range(len(routes)):
            for j in range(routes[i].flow):
                if idx < len(agents):
                    assigned_routes[idx] = i
                    idx += 1
        return assigned_routes

    def simulation_step():
        assigned_routes = assign_routes(routes, agents)

        for i in range(len(agents)):
            agents[i].add_time(routes[assigned_routes[i]].time, mean_time)

        agent_history = [0] * len(agents)

        for i in range(len(agents)):
            agent_history[agents[i].id] = assigned_routes[i]
        if debug:
            for agent in agents:
                print(f"Agent {agent.id} has time {agent.sum_time}")
        step_mean = sum(agent.sum_time for agent in agents) / len(agents)
        step_variance = sum((agent.sum_time - step_mean) ** 2 for agent in agents) / len(agents)
        step_fairness = sum(pow(agent.sum_deviation, 2) for agent in agents) / (len(agents))
        history.append(agent_history)
        mean.append(step_mean)
        variance.append(step_variance)
        fairness.append(step_fairness)
        fairness_norm.append(step_fairness /fairness[0]) # normalize fairness to make it comparable
        if(len(mean) != len(variance)):
            print("Error: mean and variance are not the same length")
            print("Mean: ", len(mean))
            print("Variance: ", len(variance))
        agents.sort()

    def check_convergence(eps, cntr):
        min_time = min(agents, key=lambda a: a.sum_time).sum_time
        max_time = max(agents, key=lambda a: a.sum_time).sum_time
        max_diff = max_time - min_time
        if max_diff > eps:
            if debug:
                print(f"Not converged yet! Max diff: {max_diff}")
            return False
        if debug:
            print(f"Converged in {cntr} steps!")
        return True
    
    def norm_fairness_convergence(eps):
        # if fairness is less than eps
        if fairness_norm[-1] < eps:
            return True
        return False
    
    # Prepare and run the simulation
    prepare_routes()
    cntr = 0
    converged = False
    convergence_iter = -1
    mean_time = sum(route.flow * route.time for route in routes) / num_agents
    while cntr < max_steps:
        cntr += 1
        simulation_step()
        if cntr % 100 == 0 and debug:
            print(f"Step {cntr} stats:")
            for agent in agents:
                print(f"Agent {agent.id} has time {agent.sum_time}")
            
        for i, threshold in enumerate(thresholds):
            if almost_convergence[i] == -1 and norm_fairness_convergence(threshold):
                almost_convergence[i] = cntr

        if check_convergence(eps, cntr):
            if not converged:
                convergence_iter = cntr
            converged = True

    # Return the simulation results
    return {
        "convergence": (converged, convergence_iter),
        "history": history,
        "mean": mean,
        "variance": variance,
        "fairness": fairness,
        "fairness_norm": fairness_norm,
        "almost_convergence": almost_convergence
    }

# Example usage of the run_simulation function
if __name__ == "__main__":
    num_agents = 9
    route_data = [(2, 15.0), (3, 14.0), (4, 9.0)]
    avg_time = sum(q * t for q, t in route_data) / num_agents
    thresholds = [0.3, 0.1, 0.05]
    results = run_simulation(num_agents, route_data, thresholds,max_steps=10000, eps=0.1)

    # Output the results
    print("Simulation Results:")
    print("History:", results["history"])
    print("Mean:", results["mean"])
    print("Variance:", results["variance"])
    print("Fairness:", results["fairness"])
    print("Converged:", results["convergence"][0])  
    print("Convergence thresholds:", results["almost_convergence"])
    print("average time", avg_time)